charms/worker/k8s/src/certificates.py

```diff
@@ -0,0 +1,310 @@
+
+    @property
+    def events(self) -> List[ops.CharmEvents]:
+        """Return the events that the Certificates library emits."""
+        return (
+            [self.certificates.on.certificate_available]
+            if self.using_external_certificates
+            else []
+        )
+
+    @property
+    def _certificate_requests(self) -> List[CertificateRequestAttributes]:
+        """Generate certificates requests for all supported components."""
+        return list(self._certificates_request_mapping.values())
+
+    @property
+    def _certificates_request_mapping(
+        self,
+    ) -> Dict[str, CertificateRequestAttributes]:
+        node_name = self.charm.get_node_name()
+        sans_ip = {"127.0.0.1", "::1"}
+        sans_dns = {
+            node_name,
+            "kubernetes",
+            "kubernetes.default",
+            "kubernetes.default.svc",
+            "kubernetes.default.svc.cluster.local",
+        }
+
+        if self.charm.is_control_plane:
+            sans_ip.update(self._get_service_ips())
+
+        extra_ips, extra_dns = self.charm.get_sorted_sans()
+        sans_ip.update(extra_ips)
+        sans_dns.update(extra_dns)
+
+        csr_attributes = {
+            "admin": CertificateRequestAttributes(
+                common_name="kubernetes:admin",
+                organization="system:masters",
+            ),
+            "controller": CertificateRequestAttributes(
+                common_name="system:kube-controller-manager",
+            ),
+            "scheduler": CertificateRequestAttributes(
+                common_name="system:kube-scheduler",
+            ),
+            "proxy": CertificateRequestAttributes(
+                common_name="system:kube-proxy",
+            ),
+            "apiserver": CertificateRequestAttributes(
+                common_name="kube-apiserver",
+                sans_dns=frozenset(sans_dns),
+                sans_ip=frozenset(sans_ip),
+            ),
+            "front-proxy-client": CertificateRequestAttributes(
+                common_name="front-proxy-client",
+            ),
+            "kubelet": CertificateRequestAttributes(
+                common_name=f"system:node:{node_name}",
+                organization="system:nodes",
+                sans_dns=frozenset(sans_dns),
+                sans_ip=frozenset(sans_ip),
+            ),
+            "kubelet-client": CertificateRequestAttributes(
+                common_name=f"system:node:{node_name}",
+                organization="system:nodes",
+            ),
+            "apiserver-kubelet-client": CertificateRequestAttributes(
+                common_name=f"system:node:{node_name}",
+                organization="system:nodes",
+            ),
+        }
+
+        if self.charm.lead_control_plane:
+            return csr_attributes
+        if self.charm.is_control_plane:
+            return {k: v for k, v in csr_attributes.items() if k in CONTROL_PLANE_CERTIFICATES}
```

Comment:
what's the situation where this code runs on the worker nodes and you need to filter out control planes here?